scripts/katana_version/__main__.py

```diff
@@ -189,35 +224,118 @@ def provenance_subcommand(args):
+            )
+            for k, v in values.items()
+        )
+        + args.suffix,
+        end="",
+    )
 
 
 def setup_provenance_subcommand(subparsers):
     parser = subparsers.add_parser(
         "provenance", help="Prints a provenance description for inclusion in artifacts. This is not a version.",
     )
 
+    class SetFormatAction(argparse.Action):
+        def __init__(self, option_strings, dest=None, nargs=None, **kwargs):
+            assert not nargs
+            self.format = None
+            self.prefix = ""
+            self.suffix = ""
+            self.separator = ""
+            self.__dict__.update(kwargs)
+            super().__init__(option_strings, dest, nargs=0)
+
+        def __call__(self, parser, namespace, values, option_string=None):
+            assert not values
+            setattr(namespace, "format", self.format)
+            setattr(namespace, "prefix", self.prefix)
+            setattr(namespace, "suffix", self.suffix)
+            setattr(namespace, "separator", self.separator)
+
     group_format = parser.add_mutually_exclusive_group()
     group_format.add_argument(
-        "--define", help="Format as #defines.", dest="format", action="store_const", const='#define {K} "{v}" \n'
+        "--define", help="Format as #defines.", action=SetFormatAction, format="#define {K} {v_double_quoted}\n",
     )
     group_format.add_argument(
-        "--yaml", help="Format as YAML.", dest="format", action="store_const", const='{k}: "{v}"\n'
+        "--yaml", help="Format as YAML.", action=SetFormatAction, format="{k}: {v_double_quoted}\n"
     )
     group_format.add_argument(
-        "--python", help="Format as Python.", dest="format", action="store_const", const='{k} = "{v}"\n'
+        "--python", help="Format as Python.", action=SetFormatAction, format="{k} = {v_double_quoted}\n"
     )
     group_format.add_argument(
-        "--format", "-f", help="Provide a format string for each value. Use the source luck.", dest="format", type=str
+        "--json",
+        help="Format as JSON.",
+        action=SetFormatAction,
+        format='  "{k}": {v_double_quoted}',
+        prefix="{\n",
+        suffix="\n}\n",
+        separator=",\n",
+    )
+
+    group_format.add_argument(
+        "--format",
+        "-f",
+        help="Provide a format string for each key-value pair. Use the source luck.",
+        dest="format",
+        type=str,
+    )
+    parser.add_argument(
+        "--separator",
+        "-j",
+        help="The separator to print between key-value pairs. Use the source luck.",
+        dest="separator",
+        type=str,
+    )
+    parser.add_argument(
+        "--prefix",
+        "-p",
+        help="Prefix to print before the first key-value pair. Use the source luck.",
+        dest="prefix",
+        type=str,
+    )
+    parser.add_argument(
+        "--suffix",
+        "-s",
+        help="Suffix to print after the last key-value pair. Use the source luck.",
```

Comment:
"source luck"? I don't understand.

Reply:
I meant "Use the source, Luke." Just a joke meaning "This is not well documented, read the source if you want to know more."